src/saveload/ser.rs

```diff
@@ -1,166 +1,176 @@
+    where
+        F: FnMut(Entity) -> Option<M>,
+    {
+        Ok(self.clone())
     }
-    serseq.end()
 }
 
-/// Serialize components from specified storages via `SerializableComponent::save`
-/// of all marked entities with provided serializer.
-/// When the component gets serialized with `SerializableComponent::save` method
-/// the closure passed in `ids` arguemnt returns `None` for unmarked `Entity`.
-/// In this case `SerializableComponent::save` may perform workaround (forget about `Entity`)
-/// or fail.
-/// So the function doesn't recursively mark referenced entities.
-/// For recursive marking see `serialize_recursive`
-pub fn serialize<'a, M, E, T, S>(
-    entities: &Entities<'a>,
-    storages: &<T as Storages<'a>>::ReadStorages,
-    markers: &ReadStorage<'a, M>,
-    serializer: S,
-) -> Result<S::Ok, S::Error>
+/// A trait which allows to serialize entities and their components.
+pub trait SerializeComponents<E, M>
 where
     M: Marker,
-    E: Display,
-    T: Components<M::Identifier, E>,
-    S: Serializer,
 {
-    let mut serseq = serializer.serialize_seq(None)?;
-    let ids = |entity| -> Option<M::Identifier> { markers.get(entity).map(Marker::id) };
-    for (entity, marker) in (&**entities, &*markers).join() {
-        serseq.serialize_element(&EntityData::<M, E, T> {
-            marker: *marker,
-            components: T::save(entity, storages, &ids).map_err(ser::Error::custom)?,
-        })?;
+    /// The data representation of the components.
+    type Data: Serialize;
+
+    /// Serialize the components of a single entiy using a entity -> marker mapping.
+    fn serialize_entity<F>(&self, entity: Entity, ids: F) -> Result<Self::Data, E>
+    where
+        F: FnMut(Entity) -> Option<M>;
+
+    /// Serialize components from specified storages via `SerializableComponent::save`
+    /// of all marked entities with provided serializer.
+    /// When the component gets serialized with `SerializableComponent::save` method
+    /// the closure passed in `ids` argument returns `None` for unmarked `Entity`.
+    /// In this case `SerializableComponent::save` may perform workaround (forget about `Entity`)
+    /// or fail.
+    /// So the function doesn't recursively mark referenced entities.
+    /// For recursive marking see `serialize_recursive`
+    fn serialize<'ms, S>(
+        &self,
+        entities: &EntitiesRes,
+        markers: &ReadStorage<M>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        E: Display,
+        S: Serializer,
+    {
+        let mut serseq = serializer.serialize_seq(None)?;
+        let ids = |entity| -> Option<M> { markers.get(entity).cloned() };
+        for (entity, marker) in (&*entities, &*markers).join() {
+            serseq.serialize_element(&EntityData::<M, Self::Data> {
+                marker: marker.clone(),
+                components: self.serialize_entity(entity, &ids)
+                    .map_err(ser::Error::custom)?, // TODO: revise
```

Comment:
What does the TODO here mean? If it's still relevant, it should be made more informative.